module/zfs/arc.c

```diff
@@ -9357,8 +9435,10 @@ l2arc_feed_thread(void *unused)
 		 * held to prevent device removal.  l2arc_dev_get_next()
 		 * will grab and release l2arc_dev_mtx.
 		 */
-		if ((dev = l2arc_dev_get_next()) == NULL)
+		if ((dev = l2arc_dev_get_next()) == NULL ||
+		    dev->l2ad_spa == NULL) {
```

Comment:
Could you please explain l2ad_spa and l2ad_vdev locking semantics, and as part of that how can l2ad_spa be NULL here if we assume the locking is correct, or how places like arc_hdr_l2hdr_destroy() won't catch NULL dereference due to a race?

Reply:
Note the above `dev->l2ad_spa == NULL` check was defensive (part of development) but is not necessary since `l2arc_dev_invalid()` already checks under locks if the spa was removed for a device. `l2arc_dev_get_next()` will never return with a NULL `dev->l2ad_spa`.

As far as locking is concerned.
The `spa_namespace_lock` covers a transition of `spa->spa_is_exporting` and the removal of a spa_t
The `spa config SCL_L2ARC` lock protects against a vdev/spa from being remove while in-use
The `l2arc_dev_mtx` protects the L2 device list and a L2 device's `l2ad_spa` and `l2ad_vdev` fields

`l2arc_dev_get_next()` will hand out L2 devices and returns with the `spa config SCL_L2ARC` lock held.  There are two possible spa exceptions that `l2arc_dev_get_next()` checks for:
* spa is being removed (`dev->l2ad_spa->spa_is_exporting` = B_TRUE) -- protected by `spa_namespace_lock`
* spa was removed (`dev->l2ad_spa` = NULL) -- protected by `l2arc_dev_mtx`

This means that when an L2 device is being removed, both the `l2arc_dev_mtx` and a writer `spa config SCL_L2ARC` lock should be held to prevent any races.  Note that the latter is currently _**not true**_ and will be remedied.

Also, `arc_hdr_l2hdr_destroy()` is protected by the `dev->l2ad_mtx` lock.  The `l2ad_vdev` can be null after a pool is exported and an async ARC removal is still in progress.

As indicated above, there is a race that I now see in `l2arc_remove_vdev()`. It was holding the  `l2arc_dev_mtx` to transition `dev->l2ad_spa` to NULL but it wasn't taking the `spa config SCL_L2ARC` (as writer) to let any inflight devices' from a past `l2arc_dev_get_next()` drain.